### DISCUSSION TITLE
how to use pysqlite custom functions?

### DISCUSSION THREAD

**Discussion by luboss** [sqlalchemy/sqlalchemy#8174] [Usage Questions] [answered]
### Describe the bug

Sqlite3 has very nice feature allowing you to define functions in Python and use them in the SQL commands. This works even with sqlalchemy on top of when using 'sqlite///:memory:' as an engine but not if the database is a file (for instance 'sqlite:///./db_file'). It throws an error: "(sqlite3.OperationalError) no such function: UDF". This seems to me a regression compared to using pure sqlite3 module.

### To Reproduce

```python
## working code using pure sqlite3 module:
import sqlite3

def udf():
    return 'udf-ok'

db = sqlite3.connect('./db_file')
db.create_function('udf', 0, udf)
for row in db.execute('SELECT UDF();'):
    print(row)

## sqlalchemy version that failes:
import sqlalchemy
import pandas

def udf():
    return 'udf-ok'

engine = sqlalchemy.create_engine('sqlite:///./db_file')
conn = engine.connect()
conn.connection.create_function('udf', 0, udf)
pandas.read_sql('SELECT UDF();', engine)

## (using 'sqlite:///:memory:' instead of file works fine)
```

### Error

---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1809                     self.dialect.do_execute_no_params(
-> 1810                         cursor, statement, context
   1811                     )

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/default.py in do_execute_no_params(self, cursor, statement, context)
    734     def do_execute_no_params(self, cursor, statement, context=None):
--> 735         cursor.execute(statement)
    736

OperationalError: no such function: UDF

The above exception was the direct cause of the following exception:

OperationalError                          Traceback (most recent call last)
<ipython-input-1-8dc66f87b557> in <module>
      8 conn = engine.connect()
      9 conn.connection.create_function('udf', 0, udf)
---> 10 pandas.read_sql('SELECT UDF();', engine)

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/pandas/io/sql.py in read_sql(sql, con, index_col, coerce_float, params, parse_dates, columns, chunksize)
    513             coerce_float=coerce_float,
    514             parse_dates=parse_dates,
--> 515             chunksize=chunksize,
    516         )
    517

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/pandas/io/sql.py in read_query(self, sql, index_col, coerce_float, parse_dates, params, chunksize)
   1293         args = _convert_params(sql, params)
   1294
-> 1295         result = self.execute(*args)
   1296         columns = result.keys()
   1297

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/pandas/io/sql.py in execute(self, *args, **kwargs)
   1160         """Simple passthrough to SQLAlchemy connectable"""
   1161         return self.connectable.execution_options(no_parameters=True).execute(
-> 1162             *args, **kwargs
   1163         )
   1164

<string> in execute(self, statement, *multiparams, **params)

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/util/deprecations.py in warned(fn, *args, **kwargs)
    399         if not skip_warning:
    400             _warn_with_version(message, version, wtype, stacklevel=3)
--> 401         return fn(*args, **kwargs)
    402
    403     doc = func.__doc__ is not None and func.__doc__ or ""

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in execute(self, statement, *multiparams, **params)
   3174         """
   3175         connection = self.connect(close_with_result=True)
-> 3176         return connection.execute(statement, *multiparams, **params)
   3177
   3178     @util.deprecated_20(

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in execute(self, statement, *multiparams, **params)
   1294                 params,
   1295                 _EMPTY_EXECUTION_OPTS,
-> 1296                 future=False,
   1297             )
   1298

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _exec_driver_sql(self, statement, multiparams, params, execution_options, future)
   1600             execution_options,
   1601             statement,
-> 1602             distilled_parameters,
   1603         )
   1604

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1861         except BaseException as e:
   1862             self._handle_dbapi_exception(
-> 1863                 e, statement, parameters, cursor, context
   1864             )
   1865

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   2042             elif should_wrap:
   2043                 util.raise_(
-> 2044                     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
   2045                 )
   2046             else:

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    206
    207         try:
--> 208             raise exception
    209         finally:
    210             # credit to

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1808                 if not evt_handled:
   1809                     self.dialect.do_execute_no_params(
-> 1810                         cursor, statement, context
   1811                     )
   1812             else:

~/.local/share/virtualenvs/inventory-new-cjyamVSQ/lib/python3.6/site-packages/sqlalchemy/engine/default.py in do_execute_no_params(self, cursor, statement, context)
    733
    734     def do_execute_no_params(self, cursor, statement, context=None):
--> 735         cursor.execute(statement)
    736
    737     def is_disconnect(self, e, connection, cursor):

OperationalError: (sqlite3.OperationalError) no such function: UDF
[SQL: SELECT UDF();]
(Background on this error at: https://sqlalche.me/e/14/e3q8)

### Versions

- OS: Debian 8.11
- Python: Python 3.6.10 (Anaconda)
- SQLAlchemy: 1.4.37 (tested with 1.4.38 on windows platform too, the same results)
- Database: Sqlite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): sqlite3

**Comment by zzzeek** [comment, marked as answer]
unfortunately not documented which it should, we can accept PRs for the pysqlite docs at https://docs.sqlalchemy.org/en/14/dialects/sqlite.html


SQLAlchemy uses a connection pool that for SQLite in SQLAlchemy 1.4 is a "nullpool" that connects each time.      so you must use [PoolEvents.connect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=poolevents+connect#sqlalchemy.events.PoolEvents.connect)

```py

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import text


def udf():
    return "udf-ok"


engine = create_engine("sqlite:///./db_file")


@event.listens_for(engine, "connect")
def connect(conn, rec):
    conn.create_function("udf", 0, udf)


for i in range(5):
    with engine.connect() as conn:
        print(conn.scalar(text("SELECT UDF()")))
```

**Reply by luboss** [reply]
thank you zzzeek so much for the solution, it will greatly help me with cleaning a lot of code

**Reply by gordthompson** [reply]
This seems to work okay as well, at least with the current main branch. Am I missing some subtlety that would come back to bite us?

```python
from sqlalchemy import create_engine, text


def udf():
    return "udf-ok"


engine = create_engine("sqlite:///./db_file")
engine.raw_connection().create_function("udf", 0, udf)

for i in range(5):
    with engine.connect() as conn:
        print(conn.scalar(text("SELECT UDF()")))
```

**Reply by zzzeek** [reply]
@gordthompson main branch uses QueuePool for SQLite, 1.4 uses NullPool.  The issue still applies to main, just will not be as apparent

**Reply by gordthompson** [reply]
> we can accept PRs for the pysqlite docs

I just tested and this works for aiosqlite under 1.4:

```python
import asyncio

from sqlalchemy import event
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine


def udf():
    return "udf-ok"


engine = create_async_engine("sqlite+aiosqlite:///./db_file")


@event.listens_for(engine.sync_engine, "connect")
def connect(conn, rec):
    conn.create_function("udf", 0, udf)


async def async_main():
    for i in range(5):
        async with engine.connect() as conn:
            result = await conn.scalar(text("SELECT UDF()"))
            print(result)


asyncio.run(async_main())
```

so should we still put the main entry under pysqlite and add a similar topic under aiosqlite with just a link to the pysqlite entry? (Sort of like we did with the SSL/certificate stuff for mysql, IIRC.)

**Reply by zzzeek** [reply]
yes, I think doc for pysqlite then aiosqlite can mention it, since aiosqlite biulds on top of pysqlite things like this will be common

**Reply by gordthompson** [reply]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3961

**Comment by 10der** [comment]
```
def _iregexp(pattern: str, value: Optional[str]) -> Optional[bool]:
    if value is None:
        return None
    return re.search(pattern, value, re.IGNORECASE) is not None


@sa.event.listens_for(sa.engine.Engine, 'connect')
def sqlite_engine_connect(dbapi_conn, connection_record) -> None:
    dbapi_conn.create_function('iregexp', 2, _iregexp)
```

```
sql = select(Article).filter(Article.title.op('iregexp')(query.query))
```

```
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) near "iregexp": syntax error
[SQL: SELECT articles.id, articles.title, articles.body, articles.chapter, articles.vote_up, articles.vote_dn 
FROM articles 
WHERE lower(articles.title) iregexp ?]
[parameters: ('інформ',)]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

what's wrong here?

**Reply by gordthompson** [reply]
> what's wrong here?

Two things, I believe.

First - You are defining a custom function, not an operator, so I think you'd need to do something like this:

```
sql = sa.select(Article).filter(sa.func.iregexp("foo", Article.title))
```

Second - SQLA_2 defaults to `QueuePool` for SQLite, and IIRC that changes the way that events are handled. (Perhaps @zzzeek can explain it better.) Try

```python
engine = sa.create_engine(connection_url, poolclass=sa.NullPool)
```

**Reply by 10der** [reply]
Thank you so much. It's works!
